### PR TITLE
Added Cancel Confirmation Modal

### LINKED ISSUE #5603: Create warning modal for order status change to Cancelled
Similar to #2814, we should implement a warning modal when merchants attempt to change the order status on a successful WC Pay order to `Cancelled`. (And perhaps other statuses?)

Like with the Refunded status change, a lack of a warning modal here can sometimes lead to painful and expensive consequences for the merchant and/or us. P2 thread pbUcTB-2e0-p2

### PR DESCRIPTION
**Pull request by jessy-p**
Fixes #5603 

#### Changes proposed in this Pull Request
- Added a confirmation modal when the user chooses to change the status of Order to Cancelled.

#### Testing instructions
1. Place an order.
2. On the admin dashboard 'Edit Order' screen, select "Cancelled" from the Order Status dropdown
3. A confirmation modal should be displayed with the option to "Do Nothing" or "Cancel Order", which warns the merchant that Cancel order will not refund the customer.
4. On click of Do Nothing, the order status dropdown is reset back to its previous status, on click of Cancel Order, it is retained as Cancelled.

Additional Testing Instructions: 
- If the order is not successful, the confirmation modal should not be displayed. e.g. an unpaid order is allowed to be cancelled without showing any confirmation modal.
- Test that there are no regressions in the Refund Confirmation Modal introduced via [PR#4987](https://github.com/Automattic/woocommerce-payments/pull/4987) as that code is also refactored.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-5.7.0#test-cancel-confirmation-modal
- [x] Not applicable, added E2E tests - Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

File: client/order/cancel-confirm-modal/index.js
Comment (naman03malhotra) [inline client/order/cancel-confirm-modal/index.js]: Is it possible to use native JS function here instead of jQuery? 
It possibly can be:
```js
document.getElementById('order_status').value = originalOrderStatus;
```
Though it would be good idea to double check if we are using jQuery for legacy reasons.
cc @ismaeldcom for input

Note for Frontend Guild: Can we start the discussion to get rid of jQuery and promote using native JS, jQuery is a bloated library. Possibly, we can reduce bundle size by getting rid of that.

File: client/order/cancel-confirm-modal/index.js
Comment (naman03malhotra) [inline client/order/cancel-confirm-modal/index.js]: Maybe a good idea to store this fragment in a variable for better code readability.

File: client/order/cancel-confirm-modal/index.js
Comment (naman03malhotra) [inline client/order/cancel-confirm-modal/index.js]: Also, this one in a separate variable for better code readability.

File: client/order/index.js
Comment (naman03malhotra) [inline client/order/index.js:94]: Nice, I like how you separated it into a different function :)

File: client/order/index.js
Comment (naman03malhotra) [inline client/order/index.js:62]: Since this flow has been refactored, adding tests for the cases covered here would be nice :) This would increase the code coverage and confidence for future refactors.

File: client/order/index.js
Comment (naman03malhotra) [inline client/order/index.js]: Just a suggestion: is it possible to use an internal library dialog UI component instead of an alert? 
IMO, alerts are harsh and outdated.

**Review by naman03malhotra — COMMENTED** [review]
Overall looks great, I left some suggestions on JS and react-specific things nothing major. I haven't tested it yet, I'll have another round of review later.

File: client/order/cancel-confirm-modal/index.js
Comment (jessy-p) [reply client/order/cancel-confirm-modal/index.js]: `jQuery( '#order_status' ).val( originalOrderStatus ).change();`

can only be replaced by the following lines, let me know what you think

```
const orderStatusElement = document.getElementById( 'order_status' );
orderStatusElement.value = originalOrderStatus;
orderStatusElement.dispatchEvent( new Event( 'change' ) );
```

File: client/order/cancel-confirm-modal/index.js
Comment (jessy-p) [reply client/order/cancel-confirm-modal/index.js]: Done

File: client/order/cancel-confirm-modal/index.js
Comment (jessy-p) [reply client/order/cancel-confirm-modal/index.js]: Done

File: client/order/index.js
Comment (jessy-p) [reply client/order/index.js]: Will check, was going by the fact that existing refund errors use an alert here on order edit
https://github.com/Automattic/woocommerce/blob/trunk/plugins/woocommerce/client/legacy/js/admin/meta-boxes-order.js#L998

File: client/order/cancel-confirm-modal/index.js
Comment (ismaeldcom) [reply client/order/cancel-confirm-modal/index.js]: >Though it would be good idea to double check if we are using jQuery for legacy reasons.

Not needed anymore, we dropped IE support ~6 months ago, so we can embrace vanilla now 🙂 

>Note for Frontend Guild: Can we start the discussion to get rid of jQuery and promote using native JS, jQuery is a bloated library. Possibly, we can reduce bundle size by getting rid of that.

I'd love to 🤔 Although I'm not sure if rewriting all the working code, just to get rid of jQuery is something we should do soon. But start writing new code in vanilla sounds like a good option.

And no issue with the bundle, we're using the global jQuery included in WP.

>can only be replaced by the following lines, let me know what you think

I'd replace `getElementById ` by `querySelector` it's faster, more consistent, and allow using any selector. But it's just a nitpick 🙂

**Comment by botwoo** [comment]
### Test the build on Jurassic Ninja 

This feature is only available for A12s. Please log in into https://jurassic.ninja/specialops/ before clicking the link below:

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=add/5603-cancel-confirmation-modal) :rocket:

- Latest commit: 5f47ba3a5671ab5ccd8a32d835d2b80675e58bc4
- Build time: 2023-03-21 12:09:13 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/order/index.js
Comment (jessy-p) [reply client/order/index.js]: Added this style of error notices. Hope it is noticeable enough, it is used in other WCPay screens on errors.
image

File: client/order/cancel-confirm-modal/index.js
Comment (jessy-p) [reply client/order/cancel-confirm-modal/index.js]: Done

File: client/order/index.js
Comment (jessy-p) [reply client/order/index.js:62]: Done

File: client/order/cancel-confirm-modal/index.js
Comment (htdat) [inline client/order/cancel-confirm-modal/index.js]: Use i18n functions for these two strings?

Also, since these strings are used in two places: the message itself and the button labels? May declare `const` for them and use the new const(s) in two places.

File: client/order/refund-confirm-modal/index.js
Comment (htdat) [inline client/order/refund-confirm-modal/index.js:78]: 👍 This is a nice improvement!

File: client/order/index.js
Comment (htdat) [inline client/order/index.js:56]: WDYT if we simplify the condition of showing the cancel modal a bit? Basically, move it into a separate `if` condition, and combine all conditions into that `if`. 

```suggestion
		} 
		
		if (
			'wc-cancelled' === this.value &&
			'wc-cancelled' !== originalStatus && 
			canRefund &&
			refundAmount > 0
```

File: tests/e2e/specs/wcpay/merchant/merchant-orders-status-change.spec.js
Comment (htdat) [inline tests/e2e/specs/wcpay/merchant/merchant-orders-status-change.spec.js]: Nitpick: 
```suggestion
	describe( 'Change Status of order to Cancelled', () => {
```

File: tests/e2e/specs/wcpay/merchant/merchant-orders-status-change.spec.js
Comment (htdat) [inline tests/e2e/specs/wcpay/merchant/merchant-orders-status-change.spec.js]: Running `document.querySelectorAll('.woocommerce-Price-amount')` on any wp-admin order page, it will return multiple NodeList/Element(s) so I think we might choose a more specific selector, prefer a unique ID or a combination with it rather than a class name.

```suggestion
			const priceElement = await page.$( '#woocommerce-order-items .total .woocommerce-Price-amount' );
```

**Review by htdat — COMMENTED** [review]
I manually tested and the PR works well. There are some inline comments. After fixing them, I think it's ready to merge (except that if other people have different opinions).

File: client/order/index.js
Comment (jessy-p) [reply client/order/index.js:56]: First set of condition = Did user click on cancelled
Second set of  condition = Is this order refundable.
Also, I feel like it reads easier when scanning the file due to similarity with the previous refund modal code, and I believe there are more modals to come.

File: tests/e2e/specs/wcpay/merchant/merchant-orders-status-change.spec.js
Comment (jessy-p) [reply tests/e2e/specs/wcpay/merchant/merchant-orders-status-change.spec.js]: Fixed, copy paste error!

File: tests/e2e/specs/wcpay/merchant/merchant-orders-status-change.spec.js
Comment (jessy-p) [reply tests/e2e/specs/wcpay/merchant/merchant-orders-status-change.spec.js]: Changed

File: client/order/cancel-confirm-modal/index.js
Comment (jessy-p) [reply client/order/cancel-confirm-modal/index.js]: Done.

File: client/order/index.js
Comment (htdat) [reply client/order/index.js:56]: Thanks for providing more details. Not a blocker here for me to approve this PR.

**Review by htdat — APPROVED** [review]
LGTM! 

✅ Manually re-tested both Cancellation and Refund Modals.

**Pull request merged**
